### PR TITLE
Pagination endpoint for donation forms 

### PR DESCRIPTION
**Pull request by Anand-Issac**
- adds endpoint for pagination for donation forms 
 - enables filtering based off: name, formType ( general->request group is null, specific -> request group is not null ), status

Screen Shot 2021-07-19 at 10 45 05 PM

### CODE REVIEW

File: server/api/resolvers/donationFormResolvers.ts [hunk 1]
```diff
@@ -18,7 +18,30 @@ const donationFormQueryResolvers = {
...
-    }
+    },
+    donationFormsPage: async (_, { skip, limit, filterOptions }, __): Promise<Array<DonationFormInterface>> => {
+        const { name, formType, status } = filterOptions;
+        let filter: any = {}
+        if (name) {
+            filter.name = { $regex: "^" + name + ".*", $options: "i" };
+        } 
+        if (formType) {
+            if (formType == "GENERAL"){
```
Comment (meganniu) [inline server/api/resolvers/donationFormResolvers.ts]: Could we get an enum for form type?

```
// donationFormModal.ts

enum DonationItemType {
    SPECIFIC = "SPECIFIC",
    GENERIC = "GENERIC",
}
```
```
\\ schema.ts
enum DonationItemType {
    SPECIFIC
    GENERIC
}


input DonationFormFilterOptions {
        name: String,
        formType: DonationItemType, 
        status: DonationItemStatus
    }
```

File: server/api/schema.ts [hunk 2]
```diff
@@ -111,6 +111,12 @@ const typeDefs = gql`
     # input FilterOptions {
     #     NOT_AVAILABLE: Boolean
     # }
+    
+    input DonationFormFilterOptions {
+        name: String,
```
Comment (meganniu) [inline server/api/schema.ts]: Could we filter by request group instead of name?

File: server/api/resolvers/donationFormResolvers.ts [hunk 3]
```diff
@@ -18,7 +18,30 @@ const donationFormQueryResolvers = {
...
+            }
+            if (formType == "SPECIFIC"){
+                filter.requestGroup = { $ne: null };
+            }
+        }
+        if (status){
+            filter.status = status;
+        }
+        return DonationForm.find(filter)
+            .sort({ name: "ascending", _id: "ascending" })
```
Comment (meganniu) [inline server/api/resolvers/donationFormResolvers.ts]: Is it possible to just sort ascending by name?

**Review by meganniu — APPROVED** [review]
WOW great work @Anand-Issac!!!
it looks and works amazing 🤩🤩

left a few comments, lmk if you any questions :))

File: server/api/resolvers/donationFormResolvers.ts [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+        const { name, formType, status } = filterOptions;
+        let filter: any = {}
+        if (name) {
+            filter.name = { $regex: "^" + name + ".*", $options: "i" };
+        } 
+        if (formType) {
+            if (formType == "GENERAL"){
+                filter.requestGroup = { $eq: null };
+            }
+            if (formType == "SPECIFIC"){
```
Comment (meganniu) [inline server/api/resolvers/donationFormResolvers.ts]: nit: use strict equality
```
if (formType === "SPECIFIC")
```

File: server/api/schema.ts (hunk 2 above)
Comment (Anand-Issac) [reply server/api/schema.ts]: I feel like we could have both, they're optional anyways.

Comment (meganniu) [reply server/api/schema.ts]: Let's choose just one. If the user specifies a name and a request group that are different, then we would have to choose between them.

If we go with filtering by name, then we would need to get the request group from the db. whereas if we filter by request group id, we can avoid getting the request group from the db

Also I don't think we're adding any feature that needs us to filter generic donation forms (donation forms without request groups)

File: server/api/resolvers/donationFormResolvers.ts (hunk 4 above)
Comment (Anand-Issac) [reply server/api/resolvers/donationFormResolvers.ts]: Done

File: server/api/resolvers/donationFormResolvers.ts (hunk 3 above)
Comment (Anand-Issac) [reply server/api/resolvers/donationFormResolvers.ts]: Done

File: server/api/schema.ts (hunk 2 above)
Comment (Anand-Issac) [reply server/api/schema.ts:116]: Done

Comment (Anand-Issac) [reply server/api/schema.ts]: Done

**Pull request merged**